- Name - Prathmesh Bhimraj Bhukele
- Contact no. 8291159068
- Prathmeshbb26@gmail.com

## Exploring the Dataset

In [1]:
import pandas as pd 
import numpy as np 

In [2]:
# loading the "Data" Dataset
data = pd.read_csv("DA_data.csv")

In [3]:
data.head()

Customer_ID  Weeks  Contract_Renewal Data_Plan  Data_Usage  \
0         1001     47                 1       Yes         2.3   
1         1002     30                 1        No         0.0   
2         1003     52                 0       Yes         4.1   
3         1004     25                 1        No         0.0   
4         1005     38                 1       Yes         2.6   

   Calls_To_Customer_Care  DayMins  DayCalls  MonthlyCharge  OverageFee  \
0                       1    128.2        97           34.2       10.00   
1                       2    210.5       105           55.3       15.50   
2                       0    221.7       120           70.5        9.75   
3                       4    183.6        97           52.7       17.00   
4                       1    110.7       112           32.2       10.50   

   RoamMins Customer_Attrition Inserted_Date  
0       9.1                 No    16-05-2022  
1       0.0                Yes    16-05-2022  
2       0.0                 No    16-05-2022  
3      10.0                Yes    16-05-2022  
4       3.3                 No    16-05-2022

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer_ID             134 non-null    int64  
 1   Weeks                   134 non-null    int64  
 2   Contract_Renewal        134 non-null    int64  
 3   Data_Plan               134 non-null    object 
 4   Data_Usage              134 non-null    float64
 5   Calls_To_Customer_Care  134 non-null    int64  
 6   DayMins                 134 non-null    float64
 7   DayCalls                134 non-null    int64  
 8   MonthlyCharge           133 non-null    float64
 9   OverageFee              134 non-null    float64
 10  RoamMins                134 non-null    float64
 11  Customer_Attrition      134 non-null    object 
 12  Inserted_Date           134 non-null    object 
dtypes: float64(5), int64(5), object(3)
memory usage: 13.7+ KB


In [5]:
data.isnull()

Customer_ID  Weeks  Contract_Renewal  Data_Plan  Data_Usage  \
0          False  False             False      False       False   
1          False  False             False      False       False   
2          False  False             False      False       False   
3          False  False             False      False       False   
4          False  False             False      False       False   
..           ...    ...               ...        ...         ...   
129        False  False             False      False       False   
130        False  False             False      False       False   
131        False  False             False      False       False   
132        False  False             False      False       False   
133        False  False             False      False       False   

     Calls_To_Customer_Care  DayMins  DayCalls  MonthlyCharge  OverageFee  \
0                     False    False     False          False       False   
1                     False    False     False          False       False   
2                     False    False     False          False       False   
3                     False    False     False          False       False   
4                     False    False     False          False       False   
..                      ...      ...       ...            ...         ...   
129                   False    False     False          False       False   
130                   False    False     False          False       False   
131                   False    False     False          False       False   
132                   False    False     False          False       False   
133                   False    False     False          False       False   

     RoamMins  Customer_Attrition  Inserted_Date  
0       False               False          False  
1       False               False          False  
2       False               False          False  
3       False               False          False  
4       False               False          False  
..        ...                 ...            ...  
129     False               False          False  
130     False               False          False  
131     False               False          False  
132     False               False          False  
133     False               False          False  

[134 rows x 13 columns]

In [6]:
data.shape

(134, 13)

In [7]:
data.size

1742

In [8]:
data.describe()

Customer_ID       Weeks  Contract_Renewal  Data_Usage  \
count   134.000000  134.000000        134.000000  134.000000   
mean   1066.902985   47.007463          0.925373    2.085821   
std      37.960183    8.867866          0.263774    0.992994   
min    1001.000000   24.000000          0.000000    0.000000   
25%    1034.250000   42.000000          1.000000    1.800000   
50%    1067.500000   50.000000          1.000000    2.250000   
75%    1100.750000   53.000000          1.000000    2.675000   
max    1130.000000   64.000000          1.000000    4.200000   

       Calls_To_Customer_Care     DayMins    DayCalls  MonthlyCharge  \
count              134.000000  134.000000  134.000000     133.000000   
mean                 1.417910  191.395522  100.611940      59.697744   
std                  1.042687   29.983501   11.699824      10.593120   
min                  0.000000  110.700000   75.000000      29.800000   
25%                  1.000000  173.300000   94.000000      52.700000   
50%                  1.000000  192.350000  100.000000      60.600000   
75%                  2.000000  212.900000  110.000000      68.400000   
max                  4.000000  273.300000  126.000000      87.100000   

       OverageFee    RoamMins  
count  134.000000  134.000000  
mean    10.440299    8.326119  
std      1.469994    3.663043  
min      6.000000    0.000000  
25%      9.750000    7.525000  
50%     10.250000    9.300000  
75%     10.750000   10.675000  
max     17.000000   13.400000

In [9]:
data.isnull().sum()

Customer_ID               0
Weeks                     0
Contract_Renewal          0
Data_Plan                 0
Data_Usage                0
Calls_To_Customer_Care    0
DayMins                   0
DayCalls                  0
MonthlyCharge             1
OverageFee                0
RoamMins                  0
Customer_Attrition        0
Inserted_Date             0
dtype: int64

## Data cleaning
1. Removed the duplicated values 
2. Imputed the missing values in Monthlychange & Overagefee using mean values
3. Converted the type of Inserted_date from int to Datetime format.
4. Changed the datatype of Date_plan , Contract_Renewal, Customer_Attrition from object to boolean .
5. Changed the datatype of Data_Usage , Calls_To_Customer_Care , OverageFee , MonthlyCharge to numeric format.

In [10]:
# Removing duplicates
data = data.drop_duplicates()

In [11]:
data.shape

(130, 13)

In [12]:
data['MonthlyCharge'].fillna(data['MonthlyCharge'].mean(), inplace=True)

In [ ]:
data['Inserted_Date'] = pd.to_datetime(data['Inserted_Date'], errors='coerce')

In [14]:
data['Data_Plan'] = data['Data_Plan'].map({'Yes': True, 'No': False})

In [15]:
data['Customer_Attrition'] = data['Customer_Attrition'].map({'Yes': True, 'No': False})

In [16]:
data['Data_Usage'] = pd.to_numeric(data['Data_Usage'])  

In [17]:
data['Calls_To_Customer_Care'] = pd.to_numeric(data['Calls_To_Customer_Care'])

In [18]:
data['OverageFee'] = pd.to_numeric(data['OverageFee'])
data['MonthlyCharge'] = pd.to_numeric(data['MonthlyCharge'])

In [19]:
data['Contract_Renewal'] = data['Contract_Renewal'].map({1: True, 0: False})

In [20]:
data.shape

(130, 13)

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 130 entries, 0 to 133
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Customer_ID             130 non-null    int64         
 1   Weeks                   130 non-null    int64         
 2   Contract_Renewal        130 non-null    bool          
 3   Data_Plan               130 non-null    bool          
 4   Data_Usage              130 non-null    float64       
 5   Calls_To_Customer_Care  130 non-null    int64         
 6   DayMins                 130 non-null    float64       
 7   DayCalls                130 non-null    int64         
 8   MonthlyCharge           130 non-null    float64       
 9   OverageFee              130 non-null    float64       
 10  RoamMins                130 non-null    float64       
 11  Customer_Attrition      130 non-null    bool          
 12  Inserted_Date           130 non-null    datetime64[ns]


## Solutions 

##### 1. What is the correlation between the number of calls to customer care and customer attrition?

In [22]:
correlation_calls_attrition = data['Calls_To_Customer_Care'].corr(data['Customer_Attrition'])
print("Correlation between calls to customer care and customer attrition:", correlation_calls_attrition)

Correlation between calls to customer care and customer attrition: 0.17755811850077777


- <span style="color:Green">  The correlation between calls to customer care and customer attritions is 0.17 which shows a weak positive correlation. </span>

##### 2. Which data plan (Yes or No) has a higher average monthly charge?

In [23]:
average_monthly_charge_by_plan = data.groupby('Data_Plan')['MonthlyCharge'].mean()
higher_average_monthly_charge_plan = average_monthly_charge_by_plan.idxmax()
print("Data plan with higher average monthly charge:", higher_average_monthly_charge_plan)

Data plan with higher average monthly charge: True


- <span style="color:Green">  "True" indicates that Customers with a data plan ('Yes') have a higher average monthly charge. </span>

##### 3. Is there any correlation between customer attrition and contract renewal?

In [24]:
correlation_attrition_renewal = data['Customer_Attrition'].corr(data['Contract_Renewal'])
print("Correlation between customer attrition and contract renewal:", correlation_attrition_renewal)

Correlation between customer attrition and contract renewal: -0.15144803708370724


- <span style="color:Green">  The correlation coefficient of approximately -0.1514 between customer attrition and contract renewal suggests a weak negative correlation </span>

##### 4. Which feature(s) have the highest correlation with customer attrition?

In [25]:
correlation_matrix = data.corr()
highest_correlation_with_attrition = correlation_matrix['Customer_Attrition'].sort_values(ascending=False).head(2)
print("Feature(s) with highest correlation with customer attrition:\n", highest_correlation_with_attrition)

Feature(s) with highest correlation with customer attrition:
 Customer_Attrition    1.000000
OverageFee            0.318281
Name: Customer_Attrition, dtype: float64


- <span style="color:Green">  The feature with the highest correlation coefficient with customer attrition is 'OverageFee' with a correlation coefficient of approximately 0.318. </span>

##### 5. Is there a difference in data usage between customers who have a data plan and those who do not?

In [26]:
data_usage_difference = data.groupby('Data_Plan')['Data_Usage'].mean()
print("Difference in data usage between customers with and without a data plan:\n", data_usage_difference)

Difference in data usage between customers with and without a data plan:
 Data_Plan
False    0.000000
True     2.398214
Name: Data_Usage, dtype: float64


- <span style="color:green"> Customers with a data plan use an average of approximately 2.39 GB of data, while those without a data plan use no data on average </span>

##### 6. What is the total revenue from customers who have a data plan and used greater than 3 GB of data

In [27]:
total_revenue_data_plan_gt_3gb = (data['MonthlyCharge'] + data['OverageFee']).loc[(data['Data_Plan'] == True) & (data['Data_Usage'] > 3)].sum()
print("Total revenue from customers with a data plan and used > 3 GB of data:", total_revenue_data_plan_gt_3gb)

Total revenue from customers with a data plan and used > 3 GB of data: 1131.8


- <span style="color:Green"> The Total revenue from customers with a data plan and used more then 3 GB of data is $ 1131.8 </span>

##### 7. What % of total revenue comes from customers who do not have a data plan?

In [28]:
#total revenue from customers without a data plan
total_revenue_no_data_plan = (data['MonthlyCharge'] + data['OverageFee']).loc[data['Data_Plan'] == False].sum()
print(total_revenue_no_data_plan)

1036.25


In [29]:
# percentage of total revenue from customers without a data plan
percentage_revenue_no_data_plan = (total_revenue_no_data_plan / (total_revenue_no_data_plan + total_revenue_data_plan_gt_3gb)) * 100
print("% of total revenue from customers who do not have a data plan:", percentage_revenue_no_data_plan)

% of total revenue from customers who do not have a data plan: 47.796406909434744


- <span style="color:Green"> The percentage of total revenue from customers who do not have a data plan is approximately 47.79% and the total revenue generated from customers with no data plan is $1036.25 </span>

##### 8. What is the ratio of total revenue between customers who have a data plan and those who do not?

In [30]:
#  total revenue from customers without a data plan
total_revenue_no_data_plan = (data['MonthlyCharge'] + data['OverageFee']).loc[data['Data_Plan'] == False].sum()
print(total_revenue_no_data_plan)

1036.25


In [31]:
# Calculating total revenue from customers with a data plan
total_revenue_data_plan = (data['MonthlyCharge'] + data['OverageFee']).loc[data['Data_Plan'] == True].sum()
print(total_revenue_data_plan)

8065.36046511628


In [45]:
# Calculating the ratio of total revenuee between customers with and without a data plan
revenue_ratio = total_revenue_data_plan / total_revenue_no_data_plan
print("Ratio :", revenue_ratio)

Ratio : 7.783218784189414


- <span style="color:Green"> This ratio indicates that the total revenue from customers with a data plan is about 7.783 times higher than the total revenue from customers without a data plan. </span>

##### 9. How many customers have a renewed contract? Are customers with a data plan less likely to renew their contract vs customers with no data plan?

In [33]:
#number of customers with renewed contracts
renewed_customers = data.loc[data['Contract_Renewal'], 'Customer_ID'].nunique()

#number of customers with renewed contracts and a data plan
data_plan_renewed_customers = data.loc[data['Data_Plan'] & data['Contract_Renewal'], 'Customer_ID'].nunique()

# number of customers with renewed contracts and no data plan
no_data_plan_renewed_customers = data.loc[~data['Data_Plan'] & data['Contract_Renewal'], 'Customer_ID'].nunique()


In [34]:
# percentage of customers with a data plan who renewed their contract
percentage_data_plan_renewed = (data_plan_renewed_customers / data['Data_Plan'].sum()) * 100
# the percentage of customers with no data plan who renewed their contract
percentage_no_data_plan_renewed = (no_data_plan_renewed_customers / (~data['Data_Plan']).sum()) * 100

In [35]:
print("Number of customers with renewed contract:", renewed_customers)
print("Number of customers with a data plan who renewed their contract:", data_plan_renewed_customers)
print("Number of customers with no data plan who renewed their contract:", no_data_plan_renewed_customers)
print("% of customers with a data plan who renewed their contract:", percentage_data_plan_renewed)
print("% of customers with no data plan who renewed their contract:", percentage_no_data_plan_renewed)

Number of customers with renewed contract: 120
Number of customers with a data plan who renewed their contract: 110
Number of customers with no data plan who renewed their contract: 10
% of customers with a data plan who renewed their contract: 98.21428571428571
% of customers with no data plan who renewed their contract: 55.55555555555556


- <span style="color:Green"> Out of 120 customers with renewed contracts, 110 have a data plan (98.21% renewal rate), while only 10 do not have a data plan (55.55% renewal rate). </span>

##### 10. What is the % of Overage Fees to Total Revenue? What is this ratio for customers with no data plan, customers using 1-3 GB of data and customers using greater than 3 GB of data?

In [36]:
# Calculatign total revenue
total_revenue = data['MonthlyCharge'].sum() + data['OverageFee'].sum()
# Calculating total overage fees
total_overage_fees = data['OverageFee'].sum()
#  the percentage of overage fees to the total revenue
percentage_overage_fees_to_revenue = (total_overage_fees / total_revenue) * 100

print("% of Overage Fees to Total Revenue:", percentage_overage_fees_to_revenue)

% of Overage Fees to Total Revenue: 14.928676690874413


- <span style="color:Green"> The percentage of overage fees to total revenue is 14.93%. </span> 

##### 11. Do customers with weeks more than 50 have a lower minute per call ratio or customers with weeks between 31 and 50 ?

In [37]:
average_minute_per_call_weeks_50_plus = data['DayMins'].loc[data['Weeks'] > 50].mean() / data['DayCalls'].loc[data['Weeks'] > 50].mean()
average_minute_per_call_weeks_31_50 = data['DayMins'].loc[(data['Weeks'] >= 31) & (data['Weeks'] <= 50)].mean() / data['DayCalls'].loc[(data['Weeks'] >= 31) & (data['Weeks'] <= 50)].mean()
print(average_minute_per_call_weeks_31_50)
print(average_minute_per_call_weeks_50_plus)

1.862611717974181
1.9277768385460692


In [38]:
if average_minute_per_call_weeks_50_plus < average_minute_per_call_weeks_31_50:
    print("Customers with weeks more than 50 have a lower minute per call ratio.")
else:
    print("Customers with weeks between 31 and 50 have a lower minute per call ratio.")

Customers with weeks between 31 and 50 have a lower minute per call ratio.


- <span style="color:Green"> The customers with weeks between 31 and 50 have a lower minute per call ratio. </span>

##### 12. What is the average overage fee for customers whose contracts are more than 30 weeks old and have a data plan and have used less than 1GB of data?

In [39]:
average_overage_fee = data['OverageFee'].loc[(data['Weeks'] > 30) & (data['Data_Plan'] == 'Yes') & (data['Data_Usage'] < 1)].mean()
print("Average overage fee for eligible customers:", average_overage_fee)

Average overage fee for eligible customers: nan


- <span style="color:Green">There are no eligible customers to calculate the average overage fee. </span>

##### 13. What is the average monthly charge for customers whose contracts are more than 50 weeks old and have a data plan and have renewed their contract?

In [40]:
# Filtered the data for customers whose contracts are more than 50 weeks old, have a data plan, and have renewed their contract
filtered_customers = data[(data['Weeks'] > 50) & (data['Data_Plan']) & (data['Contract_Renewal'] == True)]

In [41]:
# Calculated the average monthly charge whose contracts are more than 50 weeks
average_monthly_charge = filtered_customers['MonthlyCharge'].mean()
print(average_monthly_charge)

62.00680122860904


- <span style="color:Green">The average monthly charge for customers whose contracts are more than 50 weeks old, have a data plan, and have renewed their contract is 62.00</span>

##### 14. What is the average roam minutes for customers whose contracts are between 31-50 weeks and have a data plan and have used greater than 3GB of data?

In [42]:
filtered_customers = data[(data['Weeks'] >= 31) & (data['Weeks'] <= 50) & (data['Data_Plan']) & (data['Data_Usage'] > 3)]

# Calculated the average roam minutes 
average_roam_minutes = filtered_customers['RoamMins'].mean()

print(average_roam_minutes)

9.955555555555556


- <span style="color:Green"> Average roam minutes for customers whose contracts are between 31-50 weeks old, have a data plan, and have used greater than 3GB of data is 9.95 </span>

##### 15. What is the average data usage for customers whose contracts are more than 30 weeks old and have renewed their contract?

In [43]:
# Filtering the Data for customers whose contracts are more than 30 weeks old and have renewed their contract
filtered_customers = data[(data['Weeks'] > 30) & (data['Contract_Renewal'] == True)]

In [44]:
# Calculating the average data usage
average_data_usage = filtered_customers['Data_Usage'].mean()
print(average_data_usage)

2.2585585585585584


- <span style="color:Green"> Average data usage for customers whose contracts are more than 30 weeks old and have renewed their contract is 2.258</span>